Library/DasharoVariablesLib/DasharoVariablesLib.c

```diff
@@ -0,0 +1,342 @@
+  CHAR16  *VarName
+  )
+{
+  VAR_INFO  VarInfo;
+
+  VarInfo = GetVariableInfo (VarName);
+  if (VarInfo.Size == 0) {
+    DEBUG ((EFI_D_VERBOSE, "%a(): Failed to look up attributes of %s.\n", __FUNCTION__, VarName));
+  }
+
+  return VarInfo.Attributes;
+}
+
+/**
+  Reset a single Dasharo EFI variable to its default value.
+
+  @param VarName  Name of the variable to reset.
+
+  @retval RETURN_SUCCESS  Successfully measured all variables.
+**/
+STATIC
+EFI_STATUS
+ResetVariable (
+  CHAR16 *VarName
+  )
+{
+  EFI_STATUS  Status;
+  VAR_INFO    VarInfo;
+
+  VarInfo = GetVariableInfo (VarName);
+  if (VarInfo.Size == 0)
+    return EFI_NOT_FOUND;
+
+  Status = gRT->SetVariable (
+      VarName,
+      &gDasharoSystemFeaturesGuid,
+      VarInfo.Attributes,
+      VarInfo.Size,
+      &VarInfo.Data
+      );
+
+  return Status;
+}
+
+/**
+  Check whether a specified variable exists and create it if it doesn't.
+
+  The variable is created with a default value.
+
+  @param VarName  Name of the variable to initialize.
+**/
+STATIC
+VOID
+InitVariable (
+  CHAR16  *VarName
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       BufferSize;
+
+  BufferSize = 0;
+  Status = gRT->GetVariable (
+      VarName,
+      &gDasharoSystemFeaturesGuid,
+      NULL,
+      &BufferSize,
+      NULL
+      );
+
+  if (Status == EFI_NOT_FOUND) {
+    Status = ResetVariable (VarName);
+    ASSERT_EFI_ERROR (Status);
+  }
+}
+
+/**
+  Creates any Dasharo-specific variables that are missing.
+
+  The variables are created with holding their default values.
+**/
+STATIC
+VOID
+InitVariablesWithDefaults (
+  VOID
+  )
+{
+  InitVariable (DASHARO_VAR_BATTERY_CONFIG);
+  InitVariable (DASHARO_VAR_BOOT_MANAGER_ENABLED);
+  InitVariable (DASHARO_VAR_ENABLE_CAMERA);
+  InitVariable (DASHARO_VAR_ENABLE_WIFI_BT);
+  InitVariable (DASHARO_VAR_FAN_CURVE_OPTION);
+  InitVariable (DASHARO_VAR_FIRMWARE_UPDATE_MODE);
+  InitVariable (DASHARO_VAR_IOMMU_CONFIG);
+  InitVariable (DASHARO_VAR_LOCK_BIOS);
+  InitVariable (DASHARO_VAR_MEMORY_PROFILE);
+  InitVariable (DASHARO_VAR_ME_MODE);
+  InitVariable (DASHARO_VAR_NETWORK_BOOT);
+  InitVariable (DASHARO_VAR_OPTION_ROM_POLICY);
+  InitVariable (DASHARO_VAR_POWER_FAILURE_STATE);
+  InitVariable (DASHARO_VAR_PS2_CONTROLLER);
+  InitVariable (DASHARO_VAR_RESIZEABLE_BARS_ENABLED);
+  InitVariable (DASHARO_VAR_SERIAL_REDIRECTION);
+  InitVariable (DASHARO_VAR_SERIAL_REDIRECTION2);
+  InitVariable (DASHARO_VAR_SLEEP_TYPE);
+  InitVariable (DASHARO_VAR_SMM_BWP);
+  InitVariable (DASHARO_VAR_USB_MASS_STORAGE);
+  InitVariable (DASHARO_VAR_USB_STACK);
+  InitVariable (DASHARO_VAR_WATCHDOG);
+  InitVariable (DASHARO_VAR_WATCHDOG_AVAILABLE);
+}
+
+EFI_STATUS
+EFIAPI
+DasharoMeasureVariables (
+  VOID
+  )
+{
+  UINTN       Idx;
+  EFI_STATUS  Status;
+  UINTN       PrefixSize;
+  VOID        *VarData;
+  UINTN       VarSize;
+  CHAR16      *VarName;
+  CHAR8       *EventData;
+
+  for (Idx = 0; Idx < ARRAY_SIZE (mMeasuredVariables); Idx++) {

```

Comment:
I think it would really be better to loop through all variables in Dasharo GUID namespace to avoid modifications of this function in the future. Not all variables are critical, but they somewhat affect functionality or security of the platform. E.g. mMeasuredVariables does not contain USB variables nor network boot variable, which are pretty much important security options from Qubes OS user perspective (and not only Qubes OS).

Also we would avoid any arbitrary decision process about variable importance and which variables to measure and which not.

Reply:
This was prompted by feedback on APU, so if all non-APU vars get included, APU configuration must be measured as well then, right?

Reply:
Changed to measure all variables including APU config: https://github.com/Dasharo/DasharoModulePkg/compare/b81b4ceb049277445f4eff249f825d5095af273e..66f25f972c2235b27e858149686066d92c7cb82d

Reply:
Yes, that's correct. Thanks for the heads up.

Reply:
Isn't there a way to use variable protocol `GetNext` to find all variables in Dasharo GUID namespace?

I would like to avoid keeping the `mAllVariables` array if possible, or at least using `mAllVariables` to measure all variables

Reply:
Enumeration of the store for measurements: https://github.com/Dasharo/DasharoModulePkg/compare/66f25f972c2235b27e858149686066d92c7cb82d..eb710ef1197f802ec1f01a01053befb179bea3fa